src/runtime_src/core/include/experimental/xrt_graph.h

```diff
@@ -26,6 +26,205 @@
 
 typedef void *xrtGraphHandle;
 
+#ifdef __cplusplus
+
+namespace xrt {
+
+/*!
+ * @class graph
+ *
+ * The graph object represents an abstraction exported by aietool matching
+ * a specified name.
+ * The graph is created by finding matching graph name in the currently
+ * loaded xclbin.
+ */
+class graph_impl;
+class graph
+{
+public:
+  /**
+   * graph() - Constructor from a device, xclbin and graph name
+   *
+   * @param device
+   *  Device on which the graph should execute
+   * @param xclbin_id
+   *  UUID of the xclbin with the graph
+   * @param name
+   *  Name of graph to construct
+   */
+  graph(const xrt::device& device, const xrt::uuid& xclbin_id, const std::string& name);
+
+  /**
+   * reset() - Reset a graph.
+   *
+   * Reset graph by disabling tiles and enable tiles reset
+   */
+  void
+  reset() const;
+
+  /**
+   * get_timestamp() - Get timestamp of a graph.
+   *
+   * @return
+   * Timestamp in AIE cycle
+   */
+  uint64_t
+  get_timestamp() const;
+
+  /**
+   * run() - Start graph execution.
+   *
+   * Start the graph execution by default. Run forever; Or run a fixed
+   * number of iterations if specified during compilation time.
+   */
+  void
+  run();
+
+  /**
+   * run() - Start graph execution.
+   *
+   * @param iterations
+   *  Number of iterations the graph should run.
+   */
+  void
+  run(uint32_t iterations);
+
+  /**
+   * wait_done() - Wait for specified milliseconds for graph to complete.
+   *
+   * @param timeout_ms
+   *  Timeout in milliseconds
+   *
+   * Wait for done status for all the tiles in graph.
+   *
+   * The current thread will block until graph run completes or timeout.
+   */
+  void
+  wait_done(int timeout_ms);
+
+  /**
+   * wait() - Wait for graph to complete.
+   *
+   * Busy wait until graph is done.
+   *
+   * The current thread will block until graph run completes.
+   */
+  void
+  wait();
+
+  /**
+   * wait() - Wait for graph to complete for specified AIE cycles.
+   *
+   * @param cycles
+   *  AIE cycles to wait since last run starts.
+   *
+   * Wait a given AIE cycle since the last graph run and then pause the
+   * graph. If graph already runs more than the given cycles, stop the
+   * graph immediately.
+   *
+   * This API with non-zero AIE cycle is for grahp that is running forever
+   * or graph that has multi-rate core(s); zero AIE cycle means busy wait
+   * until graph is done. It has the same effect as wait() API.
+   *
+   * The current thread will block until graph run completes or is paused.
+   */
+  void
+  wait(uint64_t cycles);
```

Comment:
I wonder if we need only one wait with overloaded argument types:
```
void wait() // runt to finish
void wait(std::chrono::milliseconds timeout) // run to finish or timeout
void wait(uint64_t cycles) //run to finish or run for specified cycles
```

Reply:
I have the feeling that with @stsoe we agreed on this, even having the first one redundant by using an `= 0` default argument.

Reply:
Yes indeed :-)